shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Timelocks.hs

```diff
@@ -97,108 +106,142 @@ decodeVI = RecD ValidityInterval <! From <! From
 instance FromCBOR ValidityInterval where
   fromCBOR = decode decodeVI
 
--- ================================================================
--- Timelock' and Timelock are mutually recursive. Timelock adds a
--- convenient place to hang the memoized serialized bytes by
--- being a newtype over MemoBytes. We need to supply two things.
--- 1) A FromCBOR (Annotator (Timelock' era))  instance
--- 2) Serializer code (replaces ToCBOR instances) in the Patterns for Timelock
-
-data Timelock' era
-  = Interval' !ValidityInterval
-  | Multi' !(MultiSig era)
-  | TimelockAnd' !(StrictSeq (Timelock era)) -- Note the hidden recursion of Timelock', through Timelock.
-  | TimelockOr' !(StrictSeq (Timelock era))
-  deriving (Show, Eq, Ord, Generic, NoThunks)
-
-decTimelock' :: Era era => Word -> Decode 'Open (Annotator (Timelock' era))
-decTimelock' 0 = Ann (SumD Interval' <! From)
-decTimelock' 1 = Ann (SumD Multi') <*! From
-decTimelock' 2 = Ann (SumD TimelockAnd') <*! D (sequence <$> decodeStrictSeq fromCBOR)
-decTimelock' 3 = Ann (SumD TimelockOr') <*! D (sequence <$> decodeStrictSeq fromCBOR)
-decTimelock' k = Invalid k
-
-instance Era era => FromCBOR (Annotator (Timelock' era)) where
-  fromCBOR = decode (Summands "Annotator(Timelock' era)" decTimelock')
-
-{- The decTimelok' function replaces things like this:
-instance
-  Era era =>
-  FromCBOR (Annotator (Timelock' era))
-  where
-  fromCBOR = decodeRecordSum "Timelock" $
-    \case
-      0 -> do
-        interval <- fromCBOR -- this: fromCBOR :: (Decoder s ValidityInterval)
-        pure $ (2, pure (Interval' interval)) -- Note the pure lifts from T to (Annotator T)
-        -- Possible because ValidityInterval has no memoized
-        -- structures that remember their own bytes.
-      1 -> do
-        multi <- fromCBOR
-        pure $ (2, Multi' <$> multi)
-      2 -> do
-        timelks <- sequence <$> decodeStrictSeq fromCBOR
-        pure (2, TimelockAnd' <$> timelks)
-      3 -> do
-        timelks <- sequence <$> decodeStrictSeq fromCBOR
-        pure (2, TimelockOr' <$> timelks)
-      k -> invalidKey k
--}
+-- ==================================================================
+
+data TimelockRaw era
+  = Signature !(KeyHash 'Witness (Crypto era))
+  | AllOf !(StrictSeq (Timelock era)) -- NOTE that Timelock and
```

Comment:
Do we actually want this to be `Timelock` and not `TimelockRaw`? E.g. do we actually want the bytes of the leaf nodes stored, since we only compute the hash at the top level?

Reply:
The tipping point in this decision is that if we make the branches of the AllOf constructor TimeLockRaw, then the patterns expose that TimelockRaw exists. And the purpose of the patterns is to hide this underlying type.

Reply:
Gotcha